### PR TITLE
react-router v6 stable compatibility

### PR DESCRIPTION
**Pull request by Rugvip**
## Hey, I just made a Pull Request!

Some work towards #9913 

One of the significant breaking changes between the `beta.0` release of `react-router` and v6 stable is that route paths that are prefixed with `/` are now always considered absolute routes. This means there's a need for a refactor in plenty of places, for example the entity pages. Luckily we can ship compatibility for both methods at once, which is what this does.

Another change is that the `Route` element used to just forward the `element` when rendering. It's been changed to instead throw an error when rendered, which means we need to make sure that never happens.

Will follow this with a migration of all entity page routes, create-app updates, etc. Wanted to keep this PR slim

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by Rugvip** [comment]
The breaking changes that this causes to `PermissionedRoute` turn out to be pretty extensive, so we'll need some time to figure out the way forward. Right now my idea is to switch the logic for how route discovery is done so that every routable extension is bound to the closes ancestor element with a `path` prop. That in turn requires some changes to how plugin metadata is collected for analytics, which likely has some overlap with #10257 

The reason for all of this is that `react-router` made it a hard requirement to have `Route` elements underneath `Routes`, and I don't think it's worth the confusion to go our own way on that. That means that we can't use utility route elements like `PermissionedRoute`, but instead need to allow for those utilities to wrap the routable extension within the `element` prop, which we currently don't support.

File: packages/core-app-api/src/routing/FlatRoutes.tsx [hunk 1]
```diff
@@ -59,16 +65,30 @@ export const FlatRoutes = (props: FlatRoutesProps): JSX.Element | null => {
...
         path = path?.replace(/\/\*$/, '') ?? '/';
 
+        let element = child.props.element;
+        if (!element) {
+          element = child;
+          if (!warned && process.env.NODE_ENV !== 'test') {
+            // eslint-disable-next-line no-console
+            console.warn(
+              'DEPRECATION WARNING: All elements within <FlatRoutes> must be of type <Route> with an element prop. ' +
+                'Existing usages of <Navigate key=[path] to=[to] /> should be replaced with <Route path=[path] element=<Navigate to=[to] /> />',
```
Comment (freben) [inline packages/core-app-api/src/routing/FlatRoutes.tsx:75]:
```suggestion
                'Existing usages of <Navigate key=[path] to=[to] /> should be replaced with <Route path=[path] element={<Navigate to=[to] />} />',
```

**Review by freben — APPROVED** [review]
Looks good so far! But there were a few more changes incoming regarding the route collection right?

**Comment by Rugvip** [comment]
Merged #10507 into this, which contained the additions to the routing system, even though it's not hooked up yet. Figured we finalize this in a single PR. See #10507 for more context though

**Comment by taras** [comment]
Awesome work 👏

**Pull request merged**